libursa/src/wasm/cl.rs

```diff
@@ -0,0 +1,509 @@
+/// so allow(unused) is in place for now
+#[wasm_bindgen]
+#[allow(non_snake_case, unused)]
+pub struct CredentialDefinition {
+    publicKey: CredentialPublicKey,
+    privateKey: CredentialPrivateKey,
+    keyCorrectnessProof: CredentialKeyCorrectnessProof
+}
+
+#[wasm_bindgen]
+pub struct MasterSecret(cl::MasterSecret);
+
+#[wasm_bindgen]
+impl MasterSecret {
+    pub fn new() -> Result<MasterSecret, JsValue> {
+        Ok(MasterSecret(maperr!(cl::prover::Prover::new_master_secret())))
+    }
+}
+
+#[wasm_bindgen]
+pub struct Nonce(cl::Nonce);
+
+#[wasm_bindgen]
+impl Nonce {
+    pub fn new() -> Result<Nonce, JsValue> {
+        Ok(Nonce(maperr!(cl::new_nonce())))
+    }
+}
+
+#[wasm_bindgen]
+pub struct BlindedCredentialSecrets(cl::BlindedCredentialSecrets);
+
+#[wasm_bindgen]
+pub struct CredentialSecretsBlindingFactors(cl::CredentialSecretsBlindingFactors);
+
+#[wasm_bindgen]
+pub struct BlindedCredentialSecretsCorrectnessProof(cl::BlindedCredentialSecretsCorrectnessProof);
+
+/// Convenience class for javascript. This provides a name-value pair structure
+/// instead of a tuple. The compiler complains about unused fields
+/// so allow(unused) is in place for now
+#[wasm_bindgen]
+#[allow(non_snake_case, unused)]
+pub struct ProverBlindedCredentialSecrets {
+    blindedCredentialSecrets: BlindedCredentialSecrets,
+    credentialSecretsBlindingFactors: CredentialSecretsBlindingFactors,
+    blindedCredentialSecretsCorrectnessProof: BlindedCredentialSecretsCorrectnessProof
+}
+
+#[wasm_bindgen]
+pub struct CredentialSignature(cl::CredentialSignature);
+
+#[wasm_bindgen]
+#[allow(non_snake_case)]
+impl CredentialSignature {
+    pub fn extractIndex(&self) -> Option<u32> {
+        self.0.extract_index()
+    }
+}
+
+#[wasm_bindgen]
+pub struct SignatureCorrectnessProof(cl::SignatureCorrectnessProof);
+
+#[wasm_bindgen]
+#[allow(non_snake_case)]
+pub struct IssuedCredential {
+    credentialSignature: CredentialSignature,
+    signatureCorrectnessProof: SignatureCorrectnessProof
+}
+
+/// Convenience class for javascript. This provides a name-value pair structure
+/// instead of a tuple. The compiler complains about unused fields
+/// so allow(unused) is in place for now
+#[wasm_bindgen]
+#[allow(non_snake_case, unused)]
+pub struct IssuedCredentialWithRevocation {
+    issuedCredential: IssuedCredential,
+    delta: Option<RevocationRegistryDelta>
+}
+
+#[wasm_bindgen]
+#[derive(Serialize, Deserialize)]
+pub struct Tail(cl::Tail);
+
+#[wasm_bindgen]
+#[derive(Serialize, Deserialize)]
+pub struct RevocationPublicKey(cl::RevocationKeyPublic);
+
+#[wasm_bindgen]
+pub struct RevocationPrivateKey(cl::RevocationKeyPrivate);
+
+#[wasm_bindgen]
+#[derive(Serialize, Deserialize)]
+pub struct RevocationRegistry(cl::RevocationRegistry);
+
+#[wasm_bindgen]
+pub struct RevocationRegistryDelta(cl::RevocationRegistryDelta);
+
+#[wasm_bindgen]
+#[allow(non_snake_case)]
+impl RevocationRegistryDelta {
+    pub fn fromParts(rev_reg_from: &JsValue,
```

Comment:
Is `rev_reg_from` typed as &JsValue because wasm-bindgen does not support Option?

Reply:
Some things don't play nicely with wasm like lifetimes.